src/opts/mod.rs

```diff
@@ -1137,17 +1169,15 @@ impl From<PoolConstraints> for (usize, usize) {
 #[derive(Debug, Clone, Eq, PartialEq)]
 pub struct OptsBuilder {
     opts: MysqlOpts,
-    ip_or_hostname: String,
-    tcp_port: u16,
+    address: HostPortOrUrl,
```

Comment:
The `unreachable!()` stuff in here gets a little funny. It seems like it might be cleaner to leave these as three separate fields (`ip_or_hostname: String`, `tcp_port: u16`, `resolved_ips: Vec<IpAddr>`)? Alternatively:

```rust
struct HostPort {
    host: String,
    port: u16,
    
}

struct OptsBuilder {
    opts: MysqlOpts,
    host_port: HostPort,
}

enum HostPortOrUrl {
    HostPort(HostPort),
    Url(Url),
}
```